src/main/java/ru/vk/itmo/test/viktorkorotkikh/LSMServerImpl.java

```diff
@@ -121,124 +129,233 @@ private static void sendResponseAndCloseSessionOnError(final HttpSession session
+        if (!ALLOWED_METHODS.contains(request.getMethod())) {
+            session.sendResponse(LSMConstantResponse.methodNotAllowed(request));
+            return;
+        }
+        if (request.getMethod() == METHOD_PUT && request.getBody() == null) {
+            log.info("PUT bad request: empty body");
+            session.sendResponse(LSMConstantResponse.badRequest(request));
+            return;
+        }
+
+        if (request.getHeader(REPLICA_REQUEST_HEADER) != null) {
+            final String id = request.getParameter("id=");
+            final byte[] key = id.getBytes(StandardCharsets.UTF_8);
+            final long requestTimestamp = Instant.now().toEpochMilli();
+            session.sendResponse(processLocal(request, key, id, requestTimestamp).getOriginal());
+            return;
+        }
+
+        RequestParameters requestParameters;
+        try {
+            requestParameters = getRequestParameters(request);
+        } catch (NumberFormatException e) {
+            log.info("Bad request: invalid parameters");
+            session.sendResponse(LSMConstantResponse.badRequest(request));
+            return;
+        }
         // validate id parameter
-        final String id = request.getParameter("id=");
+        final String id = requestParameters.id();
         if (id == null || id.isEmpty()) {
             log.info("Bad request: empty id parameter");
             session.sendResponse(LSMConstantResponse.badRequest(request));
             return;
         }
-        final byte[] key = id.getBytes(StandardCharsets.UTF_8);
-        final String server = consistentHashingManager.getServerByKey(key);
-        final boolean isClusterRequest = !server.equals(selfUrl);
-
-        if (isClusterRequest) {
-            HttpRequest.Builder builder = HttpRequest.newBuilder()
-                    .uri(URI.create(server + ENTITY_V0_PATH_WITH_ID_PARAM + id))
-                    .header("cluster-url", selfUrl);
-            switch (request.getMethod()) {
-                case METHOD_GET -> builder.GET();
-                case METHOD_PUT -> {
-                    final byte[] body = request.getBody();
-                    if (body == null) {
-                        log.info("PUT bad request: empty body");
-                        session.sendResponse(LSMConstantResponse.badRequest(request));
-                        return;
-                    }
-                    builder.PUT(HttpRequest.BodyPublishers.ofByteArray(body));
-                }
-                case METHOD_DELETE -> builder.DELETE();
-                default -> {
-                    session.sendResponse(LSMConstantResponse.methodNotAllowed(request));
-                    return;
-                }
-            }
-            sendClusterRequest(request, builder.build(), session);
+        Integer ack = requestParameters.ack();
+        Integer from = requestParameters.from();
+        if (ack == null ^ from == null) {
```

Comment:
Исключающее или выглядит очень неидиоматично -- лучше писать явно:
```suggestion
        if (ack == null && from != null || ack != null && from == null) {
```